src/robot/GradientDescentIk.cpp

```diff
@@ -0,0 +1,61 @@
+#include "aikido/robot/GradientDescentIk.hpp"
+
+#include "aikido/constraint/dart/JointStateSpaceHelpers.hpp"
+#include "aikido/statespace/dart/MetaSkeletonStateSaver.hpp"
+
+using aikido::constraint::SampleGenerator;
+using aikido::constraint::dart::createSampleableBounds;
+using aikido::statespace::dart::MetaSkeletonStateSaver;
+
+namespace aikido {
+namespace robot {
+
+//==============================================================================
+GradientDescentIk::GradientDescentIk(
+    ::dart::dynamics::MetaSkeletonPtr arm,
+    ::dart::dynamics::BodyNodePtr endEffector,
+    statespace::dart::ConstMetaSkeletonStateSpacePtr armSpace,
+    common::RNG* rng)
+  : mArm(arm)
+  , mEndEffector(endEffector)
+  , mArmSpace(armSpace)
+  , mSeedConfigSampleable(
+        createSampleableBounds(mArmSpace, std::move(cloneRNGFrom(*rng)[0])))
+  , mDartIkSolver(::dart::dynamics::InverseKinematics::create(mEndEffector))
+{
+  // Do nothing.
+}
+
+//==============================================================================
+std::vector<Eigen::VectorXd> GradientDescentIk::getSolutions(
+    const Eigen::Isometry3d& targetPose, const size_t maxSolutions) const
+{
+  auto saver = MetaSkeletonStateSaver(
+      mArm, MetaSkeletonStateSaver::Options::POSITIONS);
+  DART_UNUSED(saver);
+
+  std::vector<Eigen::VectorXd> solutions;
+
+  std::shared_ptr<SampleGenerator> ikSeedGenerator
+      = mSeedConfigSampleable->createSampleGenerator();
+  auto seedState = mArmSpace->createState();
+
+  for (size_t i = 0; i < maxSolutions; i++)
+  {
+    // TODO: What to do if we can't sample a seed config?
+    if (!ikSeedGenerator->sample(seedState))
+      continue;
+
+    mArmSpace->setState(mArm.get(), seedState);
+    mDartIkSolver->getTarget()->setTransform(targetPose);
+    bool withinTol = mDartIkSolver->solveAndApply(true);
```

Comment:
I believe we should call `solve` instead of `solveAndApply` (both functions can be found [in this file](https://github.com/dartsim/dart/blob/7e3db63a07f7fcabd577efbd0487c6cd99e31c34/dart/dynamics/InverseKinematics.cpp#L133)). Because given that we are calling this in the loop, it seems wasteful to apply the solution each time. 

With that said, since we are returning more than one solutions, there is a question of which solution should get applied. I think either the first one or the last one makes sense, but regardless I think we should apply outside the loop. (And whichever solution we decide to apply, we should mention that in the documentation)

Reply:
Agreed, changed over to `solve()`.

Also which solution to apply doesn't actually matter- we're using a state saver here anyway at the beginning, which restores the metaskeleton positions when it gets destructed.